proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,353 @@
+allowed::
+
+  zero = $(id [| 0 |])
+
+If ``NoImplicitPrelude`` is enabled then you have to import ``Prelude`` as a splice
+module as well in order to use names from ``Prelude`` in negative level splices::
+
+  {-# LANGUAGE TemplateHaskell #-}
+  {-# LANGUAGE ExplicitSpliceImports #-}
+  {-# LANGUAGE NoImplicitPrelude #-}
+
+  import splice Prelude
+
+  -- accepted
+  foo = $(id [|"foo"|])
+
+  -- rejected
+  foo = id $([|"foo"|])
+
+All exported names are at level 0. Splice imports can't be rexported, unless
+they are also imported normally.
+Allowing splice imports to be exported would turn a build-time only import into a runtime
+export. Maintaining the distinction between things only needed at build-time and
+things only needed at runtime allows project dependencies to be separated in the
+same way. This is important for cross-compilation.
+
+
+
+Drawbacks
+---------
+
+* The user has to be aware of the significance of using splice imports.
+
+
+
+Alternatives
+------------
+
+* ``splice`` imports could also bring identifiers into scope so that they
+  can be used everywhere in a module, not **only** in top-level splices as
+  the proposal suggest. This approach is not taken because it means that
+  build-time only dependencies can't be distinguished from runtime dependencies
+
+* Using a pragma rather than a syntactic modifier would fit in better with
+  how ``SOURCE`` imports work and make writing backwards compatible code easier::
+
+    import {-# SPLICE #-} B
+
+* It might be proposed that an alternative would be to work out which modules
+  need to be compiled based on usage inside a module. This would compromise the
+  principle that we can learn about what's needed for a module just by looking
+  at the import list in the module header.
+
+* The extension could only apply to **home** modules, because the benefits of
+  splice imports are when using GHC's ``--make`` mode. As the proposal stands,
+  for uniformity, any module used inside a top-level splice must be marked as
+  a splice module, even if it's an external module.
+
+* Another alternative would be to allow even finer grained control of splice
+  imports so that the cases of usage at levels -1 or -2 could be distinguished.
+  This could be useful in some cross-compilation situations. This is the approach
+  suggested in the `Stage Hygience for Template Haskell proposal <https://github.com/ghc-proposals/ghc-proposals/pull/243>`_.
+
+  The syntax in this proposal can be extended in a natural way to allow for this by adding an optional
+  integer component which specifies precisely what level the imported names should be allowed at::
+
+    -- Can be used at -1
+    import splice 1 A
+    -- Can be used at -2
+    import splice 2 A
+
+  Practically, by far the most common situation is 2 stages.
+
+* Since ``ExplicitSpliceImports`` is a no-op when ``TemplateHaskell`` is
```

Comment:
The formatting is off here (see the Rendered version). I think it’s related to the fact that subsequent lines are indented more than this one.